Add UTF8 validation for strings

e-kayrakli opened the pull request:
This PR adds UTF8 validation for strings. 

Resolves #14177
Resolves https://github.com/Cray/chapel-private/issues/524

In more detail, this PR:

- Adds `runtime/include/encoding/encoding-support.h` that contains
   some of the decoding logic previously contained
   in `runtime/include/qio/qio_formatted.h`

  I imagine some more encoding/decoding support can be moved to
  `runtime/include/encoding/encoding-support.h` in the future.

- Includes `runtime/include/encoding/encoding-support.h` from
  `compiler/AST/symbol.cpp` to validate string literals statically. 
  A `USR_FATAL_CONT` is called if there is an invalid string literal.

- Adds validation support in `modules/internal/String.chpl` and uses it in the
  existing factory functions.

- Adds a new factory function `chpl_createString*NV` that does
  **N**o **V**alidation on the buffer before creating a string. 

- Adds a very rudimentary test.

- Makes string factory functions that create strings from C pointers throwing.
   
Note: Currently, most of the call sites for these factory functions are enclosed
in `try!`s. Soon after merging this, most (all?) of these calls sites will
be updated to have more appropriate error handling.

Test:
- [x] standard
- [x] gasnet

gbtitus commented:
I'll defer to @bradcray and/or interested others with regard to the licensing question.  I'll note that the website reference in `utf8-decoder.h` seems to have bit-rotted; that site doesn't exist according to my browser.

If we do go this route I'd like to see some namespace cleanup in `runtime/include/encoding-support.h`.  Granted the functions are all inline static and so don't pollute the global link-time  namespaces, but a special-purpose package like this still ought to use a private namespace.  That will make the uses of the names both more informative about where they're defined and less likely to conflict with names defined in other places.  So, maybe put a `chpl_utf8dec_` prefix or some such on the table and all the functions in there?  (I don't think the `qio_` prefix is a good idea, since it implies this is part of QIO, which it's not.)

e-kayrakli commented:
> I'll defer to @bradcray and/or interested others with regard to the licensing question. I'll note that the website reference in utf8-decoder.h seems to have bit-rotted; that site doesn't exist according to my browser.

Oh wow, yeah it is gone. I have visited that website only a week or two ago, though..

> If we do go this route I'd like to see some namespace cleanup in runtime/include/encoding-support.h. Granted the functions are all inline static and so don't pollute the global link-time namespaces, but a special-purpose package like this still ought to use a private namespace. That will make the uses of the names both more informative about where they're defined and less likely to conflict with names defined in other places. So, maybe put a chpl_utf8dec_ prefix or some such on the table and all the functions in there? (I don't think the qio_ prefix is a good idea, since it implies this is part of QIO, which it's not.)

That sounds good, I'll do that. I was going to remove the `qio` prefix but I forgot apparently.

gbtitus commented:
> ... I have visited that website only a week or two ago ...

Oh!  It's probably just offline or down, then.  That's a relief.  :-)

e-kayrakli commented:
I just rebased this branch to:
- Fix a conflict in String.chpl
- Address Greg's comment by prefixing functions with `chpl_enc_` as short for "Chapel encoding support". I didn't put `utf8` in the name because some of them are for ASCII, and we may add other encodings for IO in the future.
- Keep third party stuff in `third-party/`

I updated the PR message and rerunning the correctness tests.

mppf commented on compiler/AST/symbol.cpp:
```diff
@@ -21,6 +21,7 @@
 #define __STDC_FORMAT_MACROS
 #endif
 
+#include "../../runtime/include/encoding-support.h"
```
Hmm... This is OK but as a style matter I tend to avoid paths like this in includes. (Don't know if it presents a C portability problem). How about we add `runtime/include/encoding` or something along those lines and put the .h file in there? Then the compiler Makefile would just add a -I for that path, and this would just say `#include "encoding-support.h"`.

mppf commented on compiler/AST/symbol.cpp:
```diff
@@ -1438,11 +1439,34 @@ void createInitStringLiterals() {
+  else {
+    // TODO: somehow flag that the buffer is validated to avoid revalidation
+  }
+
```
I don't think you should move this VarSymbol creating code. You should just add it after the conditional `if (s)` below. The idea would be that the `stringLiteralsHash` would only contain validated UTF-8. But, for that to make sense, you'd need to create a sibling to `stringLiteralsHash`, e.g. `bytesLiteralsHash`, and use that in `new_BytesSymbol`.

But the bottom line is that I don't think it's a good idea to make this code do anything it doesn't absolutely have to before checking the hash. Otherwise we create problems for ourselves.

mppf commented on runtime/include/encoding-support.h:
```diff
@@ -0,0 +1,139 @@
+#ifndef _ENCODING_SUPPORT_H_
+#define _ENCODING_SUPPORT_H_
+
+// WARNING: This header is used by both the compiler and the runtime.
```
I'd write it as "Note: " rather than "WARNING: "

mppf commented on runtime/include/encoding-support.h:
```diff
@@ -0,0 +1,139 @@
+  return ((intptr_t)a) - ((intptr_t)b);
+}
+
+static inline
```
These functions need documentation about what they do and what they return

mppf commented on third-party/utf8-decoder/utf8-decoder.h:
```diff
@@ -1,12 +1,20 @@
+  #endif
+#else
+    #define RESTRICT restrict
+#endif
```
AFAIK C++ has no `restrict` keyword in any form, and something like `__restrict` is compiler-specific.

Additionally your ifdefs leave RESTRICT undefined for C++ < 2011.

Additionally since this is a header file, we have to worry about symbol conflicts, so we should make it have a name like QIO_RESTRICT, e.g.

So I would recommend this:

``` c
#ifdef __cplusplus
  #define QIO_RESTRICT
#else
  #define QIO_RESTRICT restrict
#endif
```

I don't think we care so much about `restrict` in this code for the compiler so I think it's OK to leave it out for C++.

e-kayrakli commented on modules/internal/String.chpl:
```diff
@@ -464,6 +472,7 @@ module String {
   */
   inline proc createStringWithBorrowedBuffer(s: string) {
     var ret: string;
+    try! validateEncoding(s.buff, s.len);
```
Do you think that should happen for this specific overload of `createStringWithBorrowedBuffer` or other overloads as well? How about other `createString*`?

It may be right have all of these throw, but I wanted to halt as a first step. My main concern being `createStringWithBorrowedBuffer` being used in many different places such as `Path` and `FileSystem` module. For which we have a different plan to support non UTF8 strings. And I hope to do that soon. But I didn't want to add bunch of `try/catch`es or `try!`s in those places only to replace them with something else.

That being said, this is not a strong opinion and I can do that.

mppf replied:
I'd be OK having the `try!` here transitionally but if that is the plan, you should create an issue documenting the future work, or at least mention it in the PR message.

In this specific case I'd advocate for making the `createString*` methods `throw` and then updating the call sites of them with `try!` and working over time to migrate those to better handling.

e-kayrakli commented:
Running paratest after addressing the feedback. I'll update PR message soon as well.

mppf commented on modules/internal/CString.chpl:
```diff
@@ -130,28 +130,54 @@ module CString {
-    return try ((createStringWithNewBuffer(x)).strip()): t;
+  inline proc _cast(type t:chpl_anybool, x:c_string) throws {
+    var chplString: string;
+    try! {
```
Here and in many other places this PR adds `try!` to a function that `throws`, which does not make sense to me. Even if you wanted to fix that in a 2nd PR, it would probably make sense to create that PR now by looking at this one before it's merged.

e-kayrakli replied:
Generally, I tried to avoid spiraling out to handle all the errors and limit the code changes in this PR. I put `try!` in `throws` function in multiple places to make sure that I don't change what that particular function can throw. Because that may mean the call sites of those function need to be changed, and if they throw their call sites and if they throw their call sites...

(Moreover, I'd expect this particular instance to turn into a try/catch where if a decode error is caught, an illegal argument is thrown from the catch block maybe by creating a junk string and attempting to cast it to bool and letting `StringCasts.chpl` throw the error. In other words, I don't think this function should ever throw a `DecodeError`)

Another reason I didn't want to do that particularly for c_strings is because they have an uncertain future.

mppf commented on modules/internal/ChapelError.chpl:
```diff
@@ -427,12 +444,18 @@ module ChapelError {
-    const myFileS = createStringWithNewBuffer(myFileC);
+    var myFileS: string;
+    try! {
+      myFileS = createStringWithNewBuffer(myFileC);
```
This one is interesting. It's presumably possible to have a Chapel program in a filename that is not UTF-8. In that event, this function should probably just escape the non-UTF-8 characters rather than halting.

e-kayrakli replied:
Right. Escaping non UTF8 sequences is not a capability we have right now but we are planning that for other efforts. I think it'd be reasonable to use that here once we have it.

mppf commented on modules/internal/String.chpl:
```diff
@@ -675,15 +703,21 @@ module String {
       if data.locale_id != chpl_nodeID {
         if data.len <= CHPL_SHORT_STRING_SIZE {
-          return createStringWithNewBuffer(
+          try! {
```
I don't think this one needs to validate, because the string was created on the other locale and should have validated by then. You don't necessarily need to fix in this PR but I think a comment about it would be helpful.

e-kayrakli replied:
You're right. I think `chpl__deserialize` and `localize` (as you commented below) can use `chpl_createString`. I'll make that change here as it is less controversial :)

mppf commented on modules/internal/String.chpl:
```diff
@@ -777,7 +811,9 @@ module String {
       if _local || this.locale_id == chpl_nodeID {
-        return createStringWithBorrowedBuffer(this);
+        try! {
+          return createStringWithBorrowedBuffer(this);
```
Can this call chpl_createString ?

e-kayrakli replied:
No need. Because there is no need for `try!` either. Factory functions with string arguments don't validate. I'll remove the `try!` block

mppf commented on modules/packages/ZMQ.chpl:
```diff
@@ -493,7 +493,10 @@ module ZMQ {
-        var errmsg = createStringWithNewBuffer(zmq_strerror(errno));
+        var errmsg: string;
+        try! {
+          errmsg = createStringWithNewBuffer(zmq_strerror(errno));
```
In this kind of setting, it would make more sense to escape non-UTF-8 characters

e-kayrakli replied:
Right.

I also taught about using `bytes` here. For such short-lived strings that the library-writer knows to be utf8 (probably it is even ASCII in this case), using `bytes` can be a reasonable approach.

mppf commented on modules/standard/DateTime.chpl:
```diff
@@ -433,8 +433,11 @@ module DateTime {
-                                                     c_string, c_ptrTo(buf)));
+    var str: string;
+    try! {
+      str = createStringWithNewBuffer( __primitive("cast",
```
I know your PR is not about it but AFAIK this cast primitive can be replaced with just `c_ptrTo(buf):c_string`

e-kayrakli replied:
I can make that change here.

mppf commented on modules/standard/FileSystem.chpl:
```diff
@@ -593,7 +593,9 @@ proc locale.cwd(): string throws {
     // c_strings can't cross on statements.
     err = chpl_fs_cwd(tmp);
-    ret = createStringWithOwnedBuffer(tmp);
+    try! {
```
more try! in a throwing function

e-kayrakli replied:
My reasoning is not much different than what I had for c_string _cast comment I had above.

Even though this particular instance will change, it'll probably change to something that doesn't throw any error (escaping non-UTF-8 bytes etc).

From a more philosophical standpoint, what is your concern with having `try!` in a `throws` function? The user will assume that the function will never halt and whatever error occurs in it will be either handled or propagated?

mppf replied:
> From a more philosophical standpoint, what is your concern with having try! in a throws function? The user will assume that the function will never halt and whatever error occurs in it will be either handled or propagated?

yes, generally speaking we should be avoiding halts, and if the function `throws` it is obvious how not to halt.

e-kayrakli commented:
@mppf -- With all the other relevant work to this looking fine, I am ready to merge this.

Major changes since you reviewed are:
- Add and use `chpl_createStringWithNewBufferNV` and other `chpl_createStringWith*BufferNV` variations that creates string without validation
  We need non-validated versions of all types of factory functions to cleanly avoid some multilocale leaks with string serialization.
- Have the newly added `chpl_createStringWithLiteral` call `chpl_createStringWithBorrowedBufferNV`

It is still a (very near) future work to go through the error handling introduced by this PR to avoid halts as much as possible.
